File: cpp/Stack_and_Queue/0020_Valid_Parentheses.cpp
#include <string>
#include <stack>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  bool isValid(string s) {
    if (s.size() % 2 != 0) return false;
    stack<char> stk;
    for (char c: s) {
      if (c == '(') stk.push(')');
      else if (c == '[') stk.push(']');
      else if (c == '{') stk.push('}');
      else if (stk.empty() || stk.top() != c) return false;
      else stk.pop();
    }
    return stk.empty();
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        for i in s:
            if i == '(':
                stk.append(')')
            elif i == '[':
                stk.append(']')
            elif i == '{':
                stk.append('}')
            elif stk and i == stk[-1]:
                stk.pop()
            else:
                return False
        return False if stk else True


s = "()"
print(Solution().isValid(s))  # True

s = "()[]{}"
print(Solution().isValid(s))  # True

s = "(]"
print(Solution().isValid(s))  # False
